### PR TITLE
[LanguageService] Keep Roslyn/bookkeeping in sync

### PR DESCRIPTION
**Pull request by davkean**
We only update our book keeping of files after we've called Roslyn. This avoids us getting out of sync in the face of bugs.

### CODE REVIEW

**Comment by davkean** [comment]
tag @dotnet/project-system @srivatsn

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Handlers/AbstractEvaluationCommandLineHandler.cs [hunk 1]
```diff
@@ -165,19 +165,27 @@ private void ApplyChangesToContext(IComparable version, IProjectChangeDiff diffe
...
         {
             string fullPath = _project.MakeRooted(includePath);
-            if (_paths.Remove(fullPath))
+
+            // Remove from the context first so if Roslyn throws due to a bug 
+            // or other reason, that our state of the world remains consistent
+            if (_paths.Contains(fullPath))
             {
                 RemoveFromContext(fullPath);
+                Assumes.True(_paths.Remove(fullPath));
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Handlers/AbstractEvaluationCommandLineHandler.cs]: Assumes doesn't have a conditional debug attribute right? By that I mean this code will run for release?

Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Handlers/AbstractEvaluationCommandLineHandler.cs]: Yes that is correct.

Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Handlers/AbstractEvaluationCommandLineHandler.cs]: It doesn't have conditional attribute but it has debuggerstepthrough and apparently it has an attribute that this code won't be affected by servicing releases and hence can be ngened etc. Might be better to just have the return value in a variable and call Assumes.True on that.

Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Handlers/AbstractEvaluationCommandLineHandler.cs]: Not following your logic here - but I don't have a preference either way, so I split it out.

**Pull request merged**